### PR TITLE
[ADF-4787] Invalid value of number field validates to true in visibility condition

### PR DESCRIPTION
**Pull request by mcchrys**
**Please check if the PR fulfills these requirements**

> - [x] The commit message follows our [guidelines](https://github.com/Alfresco/alfresco-ng2-components/wiki/Commit-format)
> - [ ] Tests for the changes have been added (for bug fixes / features)
> - [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x")

> - [x] Bugfix
> - [ ] Feature
> - [ ] Code style update (formatting, local variables)
> - [ ] Refactoring (no functional changes, no api changes)
> - [ ] Build related changes
> - [ ] Documentation
> - [ ] Other... Please describe:

**What is the current behaviour?** (You can also link to an open issue here)
https://issues.alfresco.com/jira/browse/ADF-4787

**What is the new behaviour?**
Fractions are not supported unless the enableFractions flag is set.

**Does this PR introduce a breaking change?** (check one with "x")

> - [ ] Yes
> - [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

File: lib/core/form/services/widget-visibility.service.ts [hunk 1]
```diff
@@ -91,6 +94,19 @@ export class WidgetVisibilityService {
 
     }
 
+    private isValidFieldValue(form: FormModel, visibilityObj: WidgetVisibilityModel): boolean {
+        let isValid = true;
+        if (visibilityObj.leftType === WidgetTypeEnum.field) {
+            form.getFormFields().forEach((formField: FormFieldModel) => {
+                if (this.isSearchedField(formField, visibilityObj.leftFormFieldId)) {
```
Comment (davidcanonieto) [inline lib/core/form/services/widget-visibility.service.ts]: It's not a good practice to interrupt a forEach this way. I guess that you're trying to find the field among all the fields. 
Can we use ```form.getFormFields().find((formField: FormFieldModel) => your stuff)``` instead?

Comment (mcchrys) [reply lib/core/form/services/widget-visibility.service.ts]: Updated.

File: lib/core/form/services/widget-visibility.service.ts [hunk 2]
```diff
@@ -67,6 +67,9 @@ export class WidgetVisibilityService {
         const rightValue = this.getRightValue(form, visibilityObj);
         const actualResult = this.evaluateCondition(leftValue, rightValue, visibilityObj.operator);
 
+        if (!this.isFieldValueValid(form, visibilityObj)) {
+            return false;
```
Comment (SilviuCPopa) [inline lib/core/form/services/widget-visibility.service.ts]: Returning a value here will block the chaining condition operator. 
I think we need to find another way to fix this problem

File: lib/core/form/services/widget-visibility.service.spec.ts [hunk 3]
```diff
@@ -916,10 +906,10 @@ describe('WidgetVisibilityService', () => {
         });
 
         it('should not set null value when the field is not visibile', () => {
-            visibilityObjTest.leftFormFieldId = 'test_4';
+            visibilityObjTest.leftFormFieldId = 'FIELD_TEST';
             visibilityObjTest.operator = '!=';
-            visibilityObjTest.rightFormFieldId = 'dropdown';
-            const fakeFormField: FormFieldModel = new FormFieldModel(formTest, jsonFieldFake);
+            visibilityObjTest.rightFormFieldId = 'RIGHT_FORM_FIELD_ID';
+            const fakeFormField: FormFieldModel = new FormFieldModel(fakeFormWithField, jsonFieldFake);
```
Comment (SilviuCPopa) [inline lib/core/form/services/widget-visibility.service.spec.ts]: Replacing the mock here and in any other places will not fix the unit tests.

File: lib/core/form/services/widget-visibility.service.ts (hunk 2 above)
Comment (mcchrys) [reply lib/core/form/services/widget-visibility.service.ts]: The expected behaviour is that the visibility condition should fail, if any one field in the condition has an invalid value. In that case, it is unnecessary to proceed with evaluating the condition

File: lib/core/form/services/widget-visibility.service.spec.ts (hunk 3 above)
Comment (mcchrys) [reply lib/core/form/services/widget-visibility.service.spec.ts]: The existing formModel (formTest) hold only the _values_ property, and does not contain the  _fields_ property in the model. In this update, we check the _isValid_ property of the fields, which will not be available in the currently used formModel (formTest), which causes the tests to fail. Hence the changes.

**Comment by DenysVuika** [comment]
@mcchrys what is the state of this PR? can you please update it

**Comment by mcchrys** [comment]
> @mcchrys what is the state of this PR? can you please update it

Some changes have been suggested by Eugenio, as to where the value validity check should be done. I am working on it.

File: lib/core/form/services/widget-visibility.service.ts [hunk 4]
```diff
@@ -91,6 +94,17 @@ export class WidgetVisibilityService {
 
     }
 
+    private isValidFieldValue(formField: FormFieldModel): boolean {
+        let isValid = true;
+        isValid = formField && formField.isValid ? true : false;
+        return isValid;
+    }
+
+    private checkVisibilityValue(visibilityObj, leftValue, rightValue) {
```
Comment (eromano) [inline lib/core/form/services/widget-visibility.service.ts]: missign return type

**Comment by eromano** [comment]
@mcchrys which is the status of this PR?

**Comment by mcchrys** [comment]
> @mcchrys which is the status of this PR?

This issue fix is in need of further discussion since there are some concerns about using the form values from _formField_ object to check validity.

**Comment by mcchrys** [comment]
Solution provided in this PR: https://github.com/Alfresco/alfresco-ng2-components/pull/5255

**Pull request closed without merging**